src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/FailedProxyCache.cs

```diff
@@ -0,0 +1,146 @@
+
+            // If not failed, ready immediately.
+            if (!_failedProxies.TryGetValue(uri, out long renewTicks))
+            {
+                return Immediate;
+            }
+
+            // If we haven't reached out renew time, the proxy can't be used.
+            if (Environment.TickCount64 < renewTicks)
+            {
+                return renewTicks;
+            }
+
+            // Renew time reached, we can remove the proxy from the cache.
+            if (TryRenewProxy(uri, renewTicks))
+            {
+                return Immediate;
+            }
+
+            // Another thread updated the cache before we could remove it.
+            // We can't know if this is a removal or an update, so check again.
+            return _failedProxies.TryGetValue(uri, out renewTicks) ? renewTicks : Immediate;
+        }
+
+        /// <summary>
+        /// Sets a proxy as failed, to avoid trying it again for some time.
+        /// </summary>
+        /// <param name="uri">The URI of the proxy.</param>
+        public void SetProxyFailed(Uri uri)
+        {
+            _failedProxies[uri] = Environment.TickCount64 + FailureTimeoutInMilliseconds;
+            Cleanup();
+        }
+
+        /// <summary>
+        /// Renews a proxy prior to its period expiring. Used when all proxies are failed to renew the proxy closest to being renewed.
+        /// </summary>
+        /// <param name="uri">The <paramref name="uri"/> of the proxy to renew.</param>
+        /// <param name="renewTicks">The current renewal time for the proxy. If the value has changed from this, the proxy will not be renewed.</param>
+        public bool TryRenewProxy(Uri uri, long renewTicks)
+        {
+            var collection = (ICollection<KeyValuePair<Uri, long>>)_failedProxies;
+            return collection.Remove(new KeyValuePair<Uri, long>(uri, renewTicks));
+        }
+
+        /// <summary>
+        /// Cleans up any old proxies that should no longer be marked as failing.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void Cleanup()
+        {
+            if (_failedProxies.Count > LargeProxyConfigBoundary && Environment.TickCount64 >= Interlocked.Read(ref _nextFlushTicks))
+            {
+                CleanupHelper();
+            }
+        }
+
+        /// <summary>
+        /// Cleans up any old proxies that should no longer be marked as failing.
+        /// </summary>
+        /// <remarks>
+        /// I expect this to never be called by <see cref="Cleanup"/> in a production system. It is only needed in the case
+        /// that a system has a very large number of proxies that the PAC script cycles through. It is moderately expensive,
+        /// so it's only run periodically and is disabled until we exceed <see cref="LargeProxyConfigBoundary"/> failed proxies.
+        /// </remarks>
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private void CleanupHelper()
```

Comment:
Presumably we expect only a relatively small number of proxies to end up as failed?  Instead of looping through, storing times, etc., did you consider instead just creating a timer that will remove the proxy from the dictionary when it fires?  That seems like it'd be a lot simpler.